sled-agent/src/nexus.rs

```diff
@@ -202,3 +167,816 @@ impl ConvertInto<nexus_client::types::DatasetKind>
+                }
+            }
+
+            let outstanding_req_ready = self.outstanding_req_ready.clone();
+            self.total_put_requests_started += 1;
+            self.outstanding_request = Some(tokio::spawn(async move {
+                let res =
+                    client.sled_agent_put(&sled_id, &info).await.map(|_| info);
+                outstanding_req_ready.notify_one();
+                (NexusOp::Put, res)
+            }));
+        } else {
+            let outstanding_req_ready = self.outstanding_req_ready.clone();
+            self.total_get_requests_started += 1;
+            self.outstanding_request = Some(tokio::spawn(async move {
+                let res = client
+                    .sled_agent_get(&sled_id)
+                    .await
+                    .map(|info| info.into_inner());
+                outstanding_req_ready.notify_one();
+                (NexusOp::Get, res)
+            }));
+        }
+    }
+
+    /// Handle a reply from nexus by extracting the value from the `JoinHandle` of
+    /// the last outstanding request.
+    async fn handle_nexus_reply(&mut self) {
+        let (op, res) = match self
+            .outstanding_request
+            .take()
+            .expect("missing JoinHandle")
+            .await
+        {
+            Ok(res) => res,
+            Err(e) => {
+                error!(self.log, "Nexus request task exited prematurely: {e}");
+                return;
+            }
+        };
+        match (op, res) {
+            (NexusOp::Get, Ok(info)) => {
+                info!(
+                    self.log,
+                    "Retrieved SledAgentInfo from Nexus: {:?}", info
+                );
+                assert!(self.nexus_known_info.is_none());
+                self.nexus_known_info = Some(NexusKnownInfo::Found(info));
+                self.total_get_requests_completed += 1;
+            }
+            (NexusOp::Put, Ok(info)) => {
+                // Unwrap Safety: we must have a known value in order to have
+                // submitted a PUT request in the first place.
+                info!(
+                    self.log,
+                    "Successfully put SledAgentInfo to nexus";
+                    "old_generation" =>
+                        %self.nexus_known_info.as_ref().unwrap().generation(),
+                    "new_generation" => %info.generation,
+                );
+                self.nexus_known_info = Some(NexusKnownInfo::Found(info));
+                self.total_put_requests_completed += 1;
+            }
+            (NexusOp::Get, Err(e)) => {
+                self.total_get_requests_completed += 1;
+                if e.status() == Some(http::StatusCode::NOT_FOUND) {
+                    if self.nexus_known_info.is_none() {
+                        self.nexus_known_info = Some(NexusKnownInfo::NotFound);
+                        return;
+                    }
+                    // Assert/panic if self.nexus_known_info.is_some() ?
```

Comment:
I'm not sure I would, FWIW, out of fear of^W^W a healthy respect for the possibility that a well-meaning server-side change will cause this error code to appear in a place where we didn't expect it to (and didn't realize it would cause sled agents to panic).

Reply:
I *think* I agree. I changed this to log a warning and reset the "known" state to `None` so that sled-agent takes another lap.